src/rt/lifetime.d

```diff
@@ -281,15 +283,7 @@ bool __setArrayAllocLength(ref BlkInfo info, size_t newlength, bool isshared, co
         {
             if(isshared)
             {
-                try synchronized(typeid(ArrayAllocLengthLock))
-                {
-                    if(*length == cast(ubyte)oldlength)
-                        *length = cast(ubyte)newlength;
-                    else
-                        return false;
-                }
-                catch (Throwable t)
-                    assert(0, "Failed to synchronize.");
+                return cas(cast(shared)length, cast(ubyte)oldlength, cast(ubyte)newlength);
```

Comment:
> This is perfectly pure.

And perfectly useless. May as well use `int[]` instead of `shared(int[])`.

Effectively all memory allocation uses a global mutex, so if that isn't pure, how does `new anything` work in a pure function? I know that memory allocation isn't actually pure, but shouldn't we apply the same rules to a mutex?